Code/GraphMol/SubstructLibrary/SubstructLibrary.cpp

```diff
@@ -0,0 +1,256 @@
+//  Copyright (c) 2017, Novartis Institutes for BioMedical Research Inc.
+//  All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are
+// met:
+//
+//     * Redistributions of source code must retain the above copyright
+//       notice, this list of conditions and the following disclaimer.
+//     * Redistributions in binary form must reproduce the above
+//       copyright notice, this list of conditions and the following
+//       disclaimer in the documentation and/or other materials provided
+//       with the distribution.
+//     * Neither the name of Novartis Institutes for BioMedical Research Inc.
+//       nor the names of its contributors may be used to endorse or promote
+//       products derived from this software without specific prior written
+//       permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+// "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+// LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+// A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+// OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+// LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+// DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+// THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+#include "SubstructLibrary.h"
+#include <RDGeneral/RDThreads.h>
+#include <boost/atomic.hpp>
+#include <GraphMol/Substruct/SubstructMatch.h>
+
+namespace RDKit {
+
+struct Bits {
+  const ExplicitBitVect *queryBits;
+  const FPHolderBase *fps;
+
+  Bits(const FPHolderBase *fps, const ROMol &m) :
+      fps(fps)
+  {
+    if(fps) {
+      queryBits = fps->getQueryBits(m);
+    } else
+      queryBits = 0;
+  }
+  
+  bool check(unsigned int idx) const {
+    if(fps) {
+      return fps->passesFilter(idx, *queryBits);
+    }
+    return true;
+  }
+};
+
+
+unsigned int SubstructLibrary::addMol(const ROMol &m) {
+  unsigned int size = mols->addMol(m);
+  if(fps) {
+    unsigned int fpsize = fps->addMol(m);
+    if(size != fpsize) {
+      std::cerr << "opps:" << size << " " << fpsize << std::endl;
+    }
+    CHECK_INVARIANT(size==fpsize,
+                    "#mols different than #fingerprints in SubstructLibrary");
+  }
+  return size;
+}
+
+namespace {
+
+// end is inclusive here
```

Comment:
it's confusing, perhaps needlessly confusing, that `end` is inclusive here but not in the functions from the public API. Why not remain consistent?